### PR TITLE
Allow any mod to be explicitly disabled in world.mt

### PR DESCRIPTION
**Pull request by sofar**
Without this patch, only worldmods and modpath mods can be
disabled, but you can never disable default game mods, like
fire, tnt, bucket, bones, which are likely to be disabled
by server admins to prevent griefing.

A message that the mod is explicitly disabled is printed out. Any
other mod remains enabled by default as usual.

The alternative solutions were insufficient for people who run
multiple servers, as updating minetest_game would overwrite the
removal of default mods.

v2: typo fix, move world.mt reading out of loop, remove debug msg.

### CODE REVIEW

File: src/mods.cpp [hunk 1]
```diff
@@ -80,12 +80,15 @@ void parseModContents(ModSpec &spec)
 	}
 }
 
-std::map<std::string, ModSpec> getModsInPath(std::string path, bool part_of_modpack)
+std::map<std::string, ModSpec> getModsInPath(std::string path, bool part_of_modpack, std::string worldpath)
```
Comment (ShadowNinja) [inline src/mods.cpp]: Should be `const std::string &`.

Comment (sofar) [inline src/mods.cpp]: I'd love to see warnings during compilation... by default.

Comment (ShadowNinja) [inline src/mods.cpp]: > I'd love to see warnings during compilation... by default.

Some warnings are enabled, some are disabled.  What does that have to do with this though?

Comment (sofar) [inline src/mods.cpp]: Sorry, I thought you were seeing a compile warning since you commented on this. Is that not the case?

Am I missing something? I don't see const being used anywhere else in the way you suggest....

Comment (ShadowNinja) [inline src/mods.cpp]: The change is for efficiency.  Instead of creating a new `std::string` object and copying the data in it you just pass a pointer.

**Comment by sofar** [comment]
Updated with comments on const std::string & passing.

**Comment by sofar** [comment]
pass const std::string also into addModsInPath()... Should be ready for merge now.

**Comment by rubenwardy** [comment]
What if there is a mod with the same name in mods/ and it is disabled? It will also disable the game mod, it appears.

**Comment by sofar** [comment]
If there's a name conflict you have to solve that first, it's already problematic to have name conflicts now, as is apparent from the mod dependency resolving code.

And yes, this option explicitly disables mods, that's what it is for.

Also, note that this code doesn't change anything for existing installations, as the game does not add load_mod_xxxx lines to the config if it's a default mod only.

**Comment by rubenwardy** [comment]
It is perfectly legal for a mod in mods/ to have the same name as a mod in
game mods. The current behaviour is that the mod in mods/ overrides the mod
in game mods, when enabled.

**Comment by sofar** [comment]
And that behavior won't change, since if you have the module enabled now, then after this patch it will still be enabled.

The patch only triggers disabling of this mod once you add "load_mod_foo = false".

**Comment by sofar** [comment]
I just reread your first comment. Yes, if you have a custom mod with the same name as a built-in mod, and you had the custom mod disabled, then you'd have a very confusing situation to begin with. The solution would be to remove the custom mod, which already was disabled anyway.

This patch makes things a lot more consistent. Disabled means disabled, for all mods, custom or builtin.

**Comment by neoascetic** [comment]
I love this idea, so +1

**Comment by PilzAdam** [comment]
Users shouldn't be able to modify the game unless the game creator wants them to do so. In the default minetest_game there are minetest.conf settings to disable fire and tnt.

**Comment by neoascetic** [comment]
If this is true, the pull request should be closed. Why is it still open?

**Comment by sofar** [comment]
@PilzAdam - that makes little to no sense as an argument. All of minetest is open source, and the entire default game is implemented as separate modules. It is by nature of the word module that it is possibly optional. 

You're also suggesting that you don't want people ever to disable Fire and TNT, yet you support options in the minetest.conf that implement just that - contradicting your own argument.

Besides, can I disable the bucket or lava in minetest.conf? Saplings? Wood? Beds? Wool?

What if someone doesn't use minetest_game but a custom game, and wants to disable a mod that's part of this non-default game?

I'd say that the settings to disable fire and TNT shouldn't be settings at all, and instead people should be redirected to this patch instead, which scales much better going forward, and will be consistent for all mods without the need for custom options that need custom code handling of options and actual code inside plugins to implement.

**Comment by PilzAdam** [comment]
@sofar players are not game creators. The game designers carefully craft the experience that they want the player to have. Generally allowing players to change anything they want in the game makes it impossible for game designers to create a good game, and ultimately leads to a worse experience for players.

Having special settings to disable some parts of the game does not contradict this, since the game creator purposely added these settings, and the game is designed to work with these.

Games are separated into mods so that they are better structured for the game designers and also to make the dependency system work with changing games. It is not the intention that mods of games that users can en- or disable them at their will.

**Comment by sofar** [comment]
@pilzadam you keep saying "users" but in reality I don't think any end-user player ever modifies minetest.conf or world.mt. However, server operators are guaranteed to modify those two files, and they will certainly and always install extra mods, replace default game mods and do so expecting that they can.

Server operators are also more than knowledgeable enough to read the server logs and discover that disabling certain mods could lead to other issues.

At the bottom line of this discussion shouldn't be whether this patch violates a minetest game designers freedom of speech (your argument) but that I as a server owner or player have the damn liberty of picking and choosing any game mod that comes bundled, so that I can upgrade that whole set of bundled mods, without carrying local git patches, or doing stupid things like deleting entire parts of a git tree and having to rebase git trees (which is just a waste of anybodies time).

Bottom line: this patch makes a single thing a whole lot easier (a lot easier) for people for something they are ALREADY doing. Nobody is going to get themselves into any new trouble, and nobody's artistic liberties are violated.

Why are we even having this discussion?

**Comment by ShadowNinja** [comment]
:+1:
Just some code style issues (spacing aroud bin-ops, indentation around conditional blocks & continuation lines, comments).

**Comment by sofar** [comment]
@ShadowNinja I see - I'm a linux kernel style C programmer, so I type "if" immediately followed by a space by habit. I've fixed up the "if" clause, but TBH I'm not sure about your other remarks, as the style of the file I edited is not uniform the same everywhere and I can't find a "style" document anywhere (if there is one, I'd be happy to fix this up further if needed, of course). Same problem with your remark on comments.

**Comment by ShadowNinja** [comment]
@sofar: Here are our official code style guidelines: http://dev.minetest.net/Code_style_guidelines

File: src/mods.cpp [hunk 2]
```diff
@@ -158,8 +173,8 @@ ModConfiguration::ModConfiguration(std::string worldpath)
 	SubgameSpec gamespec = findWorldSubgame(worldpath);
 
 	// Add all game mods and all world mods
-	addModsInPath(gamespec.gamemods_path);
-	addModsInPath(worldpath + DIR_DELIM + "worldmods");
+	addModsInPath(gamespec.gamemods_path, worldpath);
+	addModsInPath(worldpath + DIR_DELIM + "worldmods", worldpath);
```
Comment (ShadowNinja) [inline src/mods.cpp]: Passing the worldpath means that the configuration is loaded twice here.  Can you load it once and just pass a `const Settings &`?

**Comment by sofar** [comment]
I'm on a 5-week leave atm, will get to rebasing and changing this once my regular schedule is back :^)

**Comment by sofar** [comment]
@ShadowNinja looking at this and passing the worldmt_settings is easy enough, but the following goes well above my knowledge of C++:

In file included from /home/sofar/git/minetest/src/server.h:29:0,
                 from /home/sofar/git/minetest/src/client/clientlauncher.cpp:23:
/home/sofar/git/minetest/src/mods.h:80:31: error: default argument missing for parameter 3 of 'std::map<std::basic_string<char>, ModSpec> getModsInPath(std::string, bool, const Settings&)'
 std::map<std::string,ModSpec> getModsInPath(std::string path, bool part_of_modpack = false, const Settings &worldmt_
                               ^
Passing NULL as default doesn't work. I have NFC how to fix that, and didn't find anything quick on stackoverflow etc. either.

**Comment by ShadowNinja** [comment]
@sofar: The issue is that you have a required option (`worldmt`) after an optional(/default) one (`part_of_modpack`).  Just switch the order or make both of them required.  Also, space after the comma in std::map's template param list please.  :-)

.

... and of course you also have a`_` instead of a `)`...  ;-)

**Comment by sofar** [comment]
Updated this patchset - sorry for the long delay.

- worldpath removed as param that's passed down, instead a settings handle is passed down.
- fixed up code style according to comments.

as usual, please let me know if I missed anything.

**Comment by 0-afflatus** [comment]
As a game creator, I'd find this a useful feature for tracking down mod problems. I frequently resort to moving mods around in the filesystem in order to test for certain problems, particularly mapgen issues, where re-creating a worldmods folder each time would be undesirable.

@PilzAdam,
> players are not game creators.

False. Both are users of minetest and there isn't any clear demarcation between the two groups. I appreciate your efforts to get everyone to stick to some kind of design structure, but this PR would save me a fair amount of hacking. As @sofar pointed out, this is open source, so I have to accept that someone may choose to run a modified version of my game and this provides a relatively clean method for doing so.

**Comment by sofar** [comment]
minetest/minetest_game#740 is an example of option proliferation that could be prevented with this patch.

**Comment by rubenwardy** [comment]
I'd approve if the setting was explicitly changed to `load_subgame_mod_modname = false`.

**Comment by sofar** [comment]
@rubenwardy that's added complexity for the admin.

still keeps it nicely generic though, but I'm not sure the corner cases really need it.

**Comment by red-001** [comment]
@sofar  could you add support for the setting to world config menu?

**Comment by Ferk** [comment]
You can sort of workaround the problem by creating an empty mod with the same name as the one you want to disable and putting it in the worldmods/ folder of the world.

Even though a world.mt flag is cleaner.

**Comment by sofar** [comment]
I stopped caring about this issue, but it really aught to be addressed, not avoided.
